staging/src/k8s.io/client-go/discovery/cached/disk/compactable_disk_cache.go

```diff
@@ -0,0 +1,236 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package disk
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/hex"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+	"time"
+
+	"golang.org/x/sync/singleflight"
+
+	"github.com/peterbourgon/diskv"
+	"k8s.io/klog/v2"
+)
+
+const oneDay = 24 * time.Hour
+
+// CompactableDiskCache implements the httpcache.Cache interface. It delegates to
+// diskv.Diskv for Cache functions (Get, Set, Delete), and occasionally removes
+// old cached files. The specified "cacheSize" is NOT a hard limit; the disk
+// cache can breach this threshold. When cache compaction occurs, it will resize
+// the disk cache below this "cacheSize".
+//
+// Trigger: disk cache compactions are triggered when the modification time
+//   of an accessed cache file is older than one day. In this case, the
```

Comment:
I know we talked about this earlier, but on reading this I see there's a pathological case where a client somehow only every reads files which aren't in the cache, causing this condition to never get hit.

What do you think about just storing a zero byte file named e.g. `lastcompactiontime` in the cache directory, check that file's modification time (instead of the file being accessed), and touching it every time we actually do a compaction?

Reply:
(this also avoids the other pathological case where we access a bunch of files and each of them triggers a compaction)

Reply:
Good idea. I've added the `lastcompationtime` zero-sized file in the cache dir. Please have a look.